File: ios/Packages/WireCore/Sources/WireCore/ModuleContextImpl.swift
import Combine
import Contracts

/// Concrete implementation of `ModuleContext`.
///
/// Provided to each module during `AppModule.initialize(context:)`. Exposes
/// Wire Core services without leaking DI internals.
public final class ModuleContextImpl: ModuleContext {
    private let _eventBus: AppEventBus
    private let _slotRegistry: SlotRegistry
    private let _tenantPublisher: AnyPublisher<TenantConfig?, Never>
    private let _rolePublisher: AnyPublisher<Role, Never>
    private let _navigateFn: (String) -> Void

    public init(
        eventBus: AppEventBus,
        slotRegistry: SlotRegistry,
        tenantPublisher: AnyPublisher<TenantConfig?, Never>,
        rolePublisher: AnyPublisher<Role, Never>,
        navigateFn: @escaping (String) -> Void
    ) {
        _eventBus = eventBus
        _slotRegistry = slotRegistry
        _tenantPublisher = tenantPublisher
        _rolePublisher = rolePublisher
        _navigateFn = navigateFn
    }

    public var tenantConfig: AnyPublisher<TenantConfig?, Never> { _tenantPublisher }
    public var currentRole: AnyPublisher<Role, Never> { _rolePublisher }
    public var eventBus: any EventBusProtocol { _eventBus }
    public var slotRegistry: any SlotRegistryProtocol { _slotRegistry }

    public func navigate(route: String) {
        _navigateFn(route)
    }
}

File: flutter/wire/lib/src/module_context_impl.dart
import 'dart:async';
import 'package:contracts/contracts.dart';
import 'event_bus.dart';
import 'slot_registry.dart';

/// Concrete implementation of [ModuleContext].
///
/// Provided to each module during [AppModule.initialize]. Exposes
/// Wire Core services without leaking DI internals.
class ModuleContextImpl implements ModuleContext {
  final AppEventBus _eventBus;
  final WireSlotRegistry _slotRegistry;
  final StreamController<Role> _roleController;
  final StreamController<TenantConfig?> _tenantController;
  final void Function(String route) _navigateFn;

  ModuleContextImpl({
    required AppEventBus eventBus,
    required WireSlotRegistry slotRegistry,
    required StreamController<Role> roleController,
    required StreamController<TenantConfig?> tenantController,
    required void Function(String route) navigateFn,
  })  : _eventBus = eventBus,
        _slotRegistry = slotRegistry,
        _roleController = roleController,
        _tenantController = tenantController,
        _navigateFn = navigateFn;

  @override
  Stream<TenantConfig?> get tenantConfig => _tenantController.stream;

  @override
  Stream<Role> get currentRole => _roleController.stream;

  @override
  AppEventBus get eventBus => _eventBus;

  @override
  WireSlotRegistry get slotRegistry => _slotRegistry;

  @override
  void navigate(String route) => _navigateFn(route);
}

